crates/bevy_render/src/view/visibility/mod.rs

```diff
@@ -68,9 +71,44 @@ impl VisibleEntities {
     }
 }
 
+#[derive(Debug, Default, Clone)]
+pub struct EntityMeshRelationships {
+    entities_with_mesh: HashMap<Handle<Mesh>, SmallVec<[Entity; 1]>>,
+    mesh_for_entity: HashMap<Entity, Handle<Mesh>>,
+}
+
+impl EntityMeshRelationships {
+    /// Register the passed `entity` has having the passed `mesh_handle`.
+    ///
+    /// Note that this list _could_ have duplicates if an entity is assigned a new mesh and
+    /// then re-assigned the old mesh. This case should be rare so, for now, we'll risk
+    /// duplicating `Aabb` cloning+assigning.
+    fn register(&mut self, entity: Entity, mesh_handle: &Handle<Mesh>) {
+        self.entities_with_mesh
+            .entry(mesh_handle.clone_weak())
+            .or_default()
+            .push(entity);
+        self.mesh_for_entity
+            .insert(entity, mesh_handle.clone_weak());
+    }
+
+    /// Deregisters the relationship between an `Entity` and `Mesh`. Used so [`update_bounds`] can
+    /// track which relationships are still active so `Aabb`s are updated correctly.
+    fn deregister(&mut self, entity: Entity) {
+        if let Some(mesh) = self.mesh_for_entity.remove(&entity) {
+            if let Some(entities) = self.entities_with_mesh.get_mut(&mesh) {
+                if let Some(idx) = entities.iter().position(|&e| e == entity) {
+                    entities.swap_remove(idx);
+                }
+            }
+        }
+    }
```

Comment:
Avoid rightward drift by using the `?` operator.
```suggestion
    fn deregister(&mut self, entity: Entity) -> Option<()> {
        let mesh = self.mesh_for_entity.remove(&entity)?;
        let entities = self.entities_with_mesh.get_mut(&mesh)?;
        let idx = entities.iter().position(|&e| e == entity)?;
        entities.swap_remove(idx);
        Some(())
    }
```

Reply:
Hmmm I don't love adding a not-so-meaningful return value to this function just to get that syntax :thinking: That being said, [ties always go to the reviewer](https://mtlynch.io/code-review-love/#12-award-all-ties-to-your-reviewer) :-) (might not even be a tie since we have a thumbs up too!)

I'll propose one alternative but I'm happy to go with `?` if we still prefer that!:
```rust
    fn deregister(&mut self, entity: usize) {
        if let Some(entities) = self
            .mesh_for_entity
            .remove(&entity)
            .and_then(|mesh| self.entities_with_mesh.get_mut(&mesh))
        {
            if let Some(idx) = entities.iter().position(|&e| e == entity) {
                entities.swap_remove(idx);
            }
        }
    }
```

Reply:
Yeah I don't know if I want `Option<()>` for "did this remove anything". I'd either choose a `bool` or `Result` (if it's an error for nothing to have happened which might be this case?) :thinking:

We could return the thing removed by `swap_remove` to be a little less janky but no consumer ever cares because it'll always be the same thing as what you passed in :shrug: 

Having the closure seems a little weird to me too but maybe just because it's not my habit to explicitly write closures :sweat_smile: I'll ponder this for a bit and then probably just give up and do one of the things you've suggested :laughing:

Reply:
Actually I've had one more thought that I wonder if you're okay with. This `EntityMeshRelationships` maintains an invariant by having its `register` method add to two collections and its `deregister` method remove from those same two collections. I propose:
1. keeping the first lookup into `self.mesh_for_entity` as an `Option` that might be `None` because we have no idea what `Entity` was just passed to us
2. having the followup `get_mut` be a `&mut self.entities_with_mesh[&mesh]` because this _should not fail_ - if there was a mesh for the entity in `self.mesh_for_entity` there should _absolutely_ be a list of entities for that mesh
3. having the `position()` check end with an `.expect` (or similar) for the same reason - if we had a mesh in `self.mesh_for_entity` then that entity also should've been in the list for its mesh in `self.entities_with_mesh`

I _slightly_ don't like this because now you have to trust the invariants to make sure it doesn't crash and you have to use the provided methods (which we could make easier by sticking it in a module) but I think I might prefer it to the position of "this can't happen but I'm checking for it anyway because I'm not confident in my system".

We could slightly improve this by having `mesh_for_entity` not store a Mesh but instead a pointer to the `Entity` in `entities_for_mesh` but I'm not sure if that's quite worth it yet.

What do you think? Did I miss a possible edge case where this could fail or do we just not want to do this out of an abundance of caution?

Reply:
It's up to you. I believe library code should panic as little as possible. It's not cool to break user code because of a small bug in your own code (I've been responsible for breaking user code a few time and it feels bad). If you feel confident in your own code, well, it's fine to use those panicking methods. But you also have to take into consideration bevy is built by a... bevy of people, and is in active development. So things are bound to change and invariant assumptions may break as a result. And you wouldn't want to panic if a bug is introduced in the future, a `warn` or `error` log should do :)

I would recommend against `unsafe` since it puts you on "cart approval track" by marking the PR automatically as controversial.

Reply:
> bevy is built by a... bevy of people

This deserves a :1st_place_medal: 

> invariant assumptions may break as a result

Yeah this is something I hate to acknowledge. I'd hate to think that, when working on `BinaryHeap`, I'd have to worry that it's invariants would be broken. That being said, I'm not so confident to assert that here since this is more complicated than `BinaryHeap` for sure!

Now I just need to figure out which not-rightward-drifting code I like the most :-)

Also do you have any preference for the `if let Some(idx)` vs `while let Some(idx)` which I just thought of in https://github.com/bevyengine/bevy/pull/4944#discussion_r895075123?

Reply:
I chose the closure syntax :-) e6c1a1f82308ba603e5fef545c59649d076dcf74